File: internal/features/sheepmatch/snapshot.go
package sheepmatch

import (
	"fmt"
	"sort"
	"strings"

	"hdd/internal/model"
	"hdd/internal/solver"
)

func snapshotFromStartResponse(start *model.StartResponse) model.SessionSnapshot {
	if start == nil {
		return model.SessionSnapshot{}
	}
	return model.SessionSnapshot{
		Difficulty: start.Difficulty,
		SessionID:  start.SessionID,
		SlotLimit:  start.SlotLimit,
		Powerups:   start.Powerups,
		Status:     start.Status,
		Tiles:      cloneTilesWithoutIDs(start.Tiles, start.Slots),
		SlotTiles:  cloneSlotTiles(start.SlotTiles, start.Slots, start.Tiles),
		MoveCount:  start.MoveCount,
	}
}

func snapshotFromHistoryItem(item *model.HistoryItem) model.SessionSnapshot {
	if item == nil {
		return model.SessionSnapshot{}
	}
	return model.SessionSnapshot{
		Difficulty: item.Difficulty,
		SessionID:  item.SessionID,
		SlotLimit:  item.SlotLimit,
		Powerups:   item.Powerups,
		Status:     item.Status,
		Tiles:      cloneTilesWithoutIDs(item.Tiles, item.Slots),
		SlotTiles:  cloneSlotTiles(item.SlotTiles, item.Slots, item.Tiles),
		MoveCount:  item.MoveCount,
	}
}

func historyItemToStartResponse(item *model.HistoryItem) *model.StartResponse {
	if item == nil {
		return nil
	}
	return &model.StartResponse{
		Difficulty: item.Difficulty,
		MoveCount:  item.MoveCount,
		Powerups:   item.Powerups,
		SessionID:  item.SessionID,
		SlotLimit:  item.SlotLimit,
		Slots:      append([]int(nil), item.Slots...),
		SlotTiles:  cloneTiles(item.SlotTiles),
		Status:     item.Status,
		Tiles:      cloneTiles(item.Tiles),
	}
}

func orderedTilesByIDDesc(tiles []model.Tile) []model.Tile {
	ordered := cloneTiles(tiles)
	sort.Slice(ordered, func(i, j int) bool {
		return ordered[i].ID > ordered[j].ID
	})
	return ordered
}

func applyPlannedActionLocally(snapshot model.SessionSnapshot, action solver.Action) (model.SessionSnapshot, error) {
	switch action.Kind {
	case "click":
		return applyLocalClickSnapshot(snapshot, action.TileID)
	case "undo", "remove", "shuffle":
		next := snapshot
		next.Status = action.Kind
		return next, nil
	case "abandon":
		next := snapshot
		next.Status = "abandoned"
		return next, nil
	default:
		return model.SessionSnapshot{}, fmt.Errorf("不支持的动作：%s", action.Kind)
	}
}

func applyLocalClickSnapshot(snapshot model.SessionSnapshot, tileID int) (model.SessionSnapshot, error) {
	ordered := orderedTilesByIDDesc(snapshot.Tiles)
	for _, tile := range ordered {
		if tile.ID != tileID {
			continue
		}
		next := snapshot
		next.Tiles = removeTileByID(snapshot.Tiles, tileID)
		next.SlotTiles = append(cloneTiles(snapshot.SlotTiles), tile)
		if len(next.SlotTiles) > next.SlotLimit {
			next.Status = "failed"
			return next, nil
		}
		patternCount := 0
		for _, slotTile := range next.SlotTiles {
			if slotTile.Pattern == tile.Pattern {
				patternCount++
			}
		}
		if patternCount >= 3 {
			filtered := make([]model.Tile, 0, len(next.SlotTiles)-3)
			removed := 0
			for _, slotTile := range next.SlotTiles {
				if slotTile.Pattern == tile.Pattern && removed < 3 {
					removed++
					continue
				}
				filtered = append(filtered, slotTile)
			}
			next.SlotTiles = filtered
		}
		next.MoveCount++
		if len(next.Tiles) == 0 && len(next.SlotTiles) == 0 {
			next.Status = "won"
		} else {
			next.Status = snapshot.Status
		}
		return next, nil
	}
	return model.SessionSnapshot{}, fmt.Errorf("目标方块已不在棋盘上")
}

func isSlotFullError(err error) bool {
	if err == nil {
		return false
	}
	message := strings.ToLower(strings.TrimSpace(err.Error()))
	return strings.Contains(message, "槽位已满") || strings.Contains(message, "slot full")
}

func snapshotFromStepResponse(previous model.SessionSnapshot, stepResp *model.StepResponse) model.SessionSnapshot {
	if stepResp == nil {
		return previous
	}
	next := previous
	next.MoveCount = stepResp.MoveCount
	next.Status = stepResp.Status
	if stepResp.SessionID != 0 {
		next.SessionID = stepResp.SessionID
	}
	if stepResp.SlotLimit > 0 {
		next.SlotLimit = stepResp.SlotLimit
	}
	next.Powerups = stepResp.Powerups
	next.SlotTiles = resolveSlotTiles(previous, stepResp)
	next.Tiles = cloneTilesWithoutIDs(stepResp.Tiles, collectTileIDs(next.SlotTiles))
	if len(stepResp.Removed) > 0 {
		next.Tiles = removeTilesByID(next.Tiles, stepResp.Removed)
	}
	return next
}

func resolveSlotTiles(previous model.SessionSnapshot, stepResp *model.StepResponse) []model.Tile {
	if len(stepResp.Slots) == 0 {
		return nil
	}
	lookup := make(map[int]model.Tile, len(previous.SlotTiles)+len(previous.Tiles)+len(stepResp.Tiles))
	for _, tile := range previous.SlotTiles {
		lookup[tile.ID] = tile
	}
	for _, tile := range previous.Tiles {
		lookup[tile.ID] = tile
	}
	for _, tile := range stepResp.Tiles {
		lookup[tile.ID] = tile
	}
	result := make([]model.Tile, 0, len(stepResp.Slots))
	for _, id := range stepResp.Slots {
		if tile, ok := lookup[id]; ok {
			result = append(result, tile)
		}
	}
	return result
}

func collectTileIDs(tiles []model.Tile) []int {
	ids := make([]int, 0, len(tiles))
	for _, tile := range tiles {
		ids = append(ids, tile.ID)
	}
	return ids
}

func cloneTilesWithoutIDs(tiles []model.Tile, excluded []int) []model.Tile {
	excludedSet := make(map[int]struct{}, len(excluded))
	for _, id := range excluded {
		excludedSet[id] = struct{}{}
	}
	result := make([]model.Tile, 0, len(tiles))
	for _, tile := range tiles {
		if _, ok := excludedSet[tile.ID]; ok {
			continue
		}
		result = append(result, tile)
	}
	return result
}

func cloneSlotTiles(explicit []model.Tile, slots []int, board []model.Tile) []model.Tile {
	if len(explicit) > 0 {
		return cloneTiles(explicit)
	}
	if len(slots) == 0 {
		return nil
	}
	lookup := make(map[int]model.Tile, len(board))
	for _, tile := range board {
		lookup[tile.ID] = tile
	}
	result := make([]model.Tile, 0, len(slots))
	for _, id := range slots {
		if tile, ok := lookup[id]; ok {
			result = append(result, tile)
		}
	}
	return result
}

func removeTileByID(tiles []model.Tile, tileID int) []model.Tile {
	result := make([]model.Tile, 0, len(tiles))
	for _, tile := range tiles {
		if tile.ID == tileID {
			continue
		}
		result = append(result, tile)
	}
	return result
}

func removeTilesByID(tiles []model.Tile, ids []int) []model.Tile {
	if len(ids) == 0 {
		return cloneTiles(tiles)
	}
	removed := make(map[int]struct{}, len(ids))
	for _, id := range ids {
		removed[id] = struct{}{}
	}
	result := make([]model.Tile, 0, len(tiles))
	for _, tile := range tiles {
		if _, ok := removed[tile.ID]; ok {
			continue
		}
		result = append(result, tile)
	}
	return result
}

func cloneTiles(tiles []model.Tile) []model.Tile {
	result := make([]model.Tile, len(tiles))
	copy(result, tiles)
	return result
}

File: src/workflows/sheepmatch/snapshot.rs
use crate::model::{HistoryItem, SessionSnapshot, StartResponse, StepResponse, Tile};

pub(super) fn snapshot_from_start_response(start: &StartResponse) -> SessionSnapshot {
    SessionSnapshot {
        difficulty: normalize_start_difficulty(start),
        session_id: start.session_id,
        slot_limit: normalize_slot_limit(start.slot_limit, 0),
        powerups: start.powerups.clone(),
        status: normalize_round_status(&start.status),
        tiles: clone_tiles_without_ids(&start.tiles, &start.slots),
        slot_tiles: clone_slot_tiles(&start.slot_tiles, &start.slots, &start.tiles),
        move_count: start.move_count,
    }
}

pub(super) fn history_item_to_start_response(item: &HistoryItem) -> StartResponse {
    StartResponse {
        difficulty: item.difficulty.clone(),
        move_count: item.move_count,
        powerups: item.powerups.clone(),
        session_id: item.session_id,
        slot_limit: item.slot_limit,
        slots: item.slots.clone(),
        slot_tiles: item.slot_tiles.clone(),
        status: item.status.clone(),
        tiles: item.tiles.clone(),
        ..StartResponse::default()
    }
}

pub(super) fn snapshot_from_step_response(
    previous: &SessionSnapshot,
    step: &StepResponse,
) -> SessionSnapshot {
    let mut next = previous.clone();
    next.move_count = step.move_count;
    next.status = if step.status.trim().is_empty() {
        previous.status.clone()
    } else {
        step.status.clone()
    };
    if step.session_id != 0 {
        next.session_id = step.session_id;
    }
    if step.slot_limit > 0 {
        next.slot_limit = step.slot_limit;
    }
    if let Some(powerups) = &step.powerups {
        next.powerups = powerups.clone();
    }
    if step.slots.is_some() {
        next.slot_tiles = resolve_slot_tiles(previous, step);
    }
    if let Some(tiles) = &step.tiles {
        next.tiles = clone_tiles_without_ids(tiles, &collect_tile_ids(&next.slot_tiles));
    }
    if !step.removed.is_empty() {
        next.tiles = remove_tiles_by_id(&next.tiles, &step.removed);
    }
    next
}

fn resolve_slot_tiles(previous: &SessionSnapshot, step: &StepResponse) -> Vec<Tile> {
    let Some(slots) = &step.slots else {
        return previous.slot_tiles.clone();
    };
    let mut lookup = std::collections::HashMap::<i32, Tile>::new();
    let step_tiles = step.tiles.as_deref().unwrap_or(&[]);
    for tile in previous
        .slot_tiles
        .iter()
        .chain(previous.tiles.iter())
        .chain(step_tiles.iter())
    {
        lookup.insert(tile.id, tile.clone());
    }
    slots
        .iter()
        .filter_map(|id| lookup.get(id).cloned())
        .collect()
}

fn normalize_start_difficulty(start: &StartResponse) -> String {
    start.difficulty.trim().to_string()
}

fn normalize_round_status(status: &str) -> String {
    let trimmed = status.trim();
    if trimmed.is_empty() {
        "pending".to_string()
    } else {
        trimmed.to_string()
    }
}

fn normalize_slot_limit(slot_limit: i32, fallback: i32) -> i32 {
    if slot_limit > 0 { slot_limit } else { fallback }
}

fn clone_tiles_without_ids(tiles: &[Tile], excluded: &[i32]) -> Vec<Tile> {
    let excluded: std::collections::HashSet<i32> = excluded.iter().copied().collect();
    tiles
        .iter()
        .filter(|tile| !excluded.contains(&tile.id))
        .cloned()
        .collect()
}

fn clone_slot_tiles(explicit: &[Tile], slots: &[i32], board: &[Tile]) -> Vec<Tile> {
    if !explicit.is_empty() {
        return explicit.to_vec();
    }
    let lookup = board
        .iter()
        .map(|tile| (tile.id, tile.clone()))
        .collect::<std::collections::HashMap<_, _>>();
    slots
        .iter()
        .filter_map(|id| lookup.get(id).cloned())
        .collect()
}

pub(super) fn collect_tile_ids(tiles: &[Tile]) -> Vec<i32> {
    tiles.iter().map(|tile| tile.id).collect()
}

pub(super) fn fixed_click_queue(snapshot: &SessionSnapshot) -> Vec<i32> {
    let mut ids = collect_tile_ids(&snapshot.tiles);
    ids.sort_unstable_by(|a, b| b.cmp(a));
    ids
}

pub(super) fn is_stale_click_error(message: &str) -> bool {
    let message = message.trim().to_ascii_lowercase();
    message.contains("目标方块已不在棋盘上")
        || message.contains("tile not on board")
        || message.contains("tile is covered")
        || message.contains("session not found")
        || message.contains("invalid action")
}

fn remove_tiles_by_id(tiles: &[Tile], ids: &[i32]) -> Vec<Tile> {
    let removed: std::collections::HashSet<i32> = ids.iter().copied().collect();
    tiles
        .iter()
        .filter(|tile| !removed.contains(&tile.id))
        .cloned()
        .collect()
}

pub(super) fn is_solved(snapshot: &SessionSnapshot) -> bool {
    snapshot.status == "won" || (snapshot.tiles.is_empty() && snapshot.slot_tiles.is_empty())
}

pub(super) fn is_slot_full_error(message: &str) -> bool {
    let message = message.trim().to_ascii_lowercase();
    message.contains("槽位已满") || message.contains("slot full")
}
